drivers/timer/nrf_rtc_timer.c

```diff
@@ -238,12 +431,56 @@ static void sys_clock_timeout_handler(int32_t chan,
 		/* protection is not needed because we are in the RTC interrupt
 		 * so it won't get preempted by the interrupt.
 		 */
-		compare_set(chan, last_count + CYC_PER_TICK,
-					  sys_clock_timeout_handler, NULL);
+		compare_set(chan, last_count + CYC_PER_TICK, sys_clock_timeout_handler, NULL);
 	}
 
-	sys_clock_announce(IS_ENABLED(CONFIG_TICKLESS_KERNEL) ?
-						dticks : (dticks > 0));
+	sys_clock_announce(IS_ENABLED(CONFIG_TICKLESS_KERNEL) ? dticks : (dticks > 0));
+}
+
+static void process_channel(int32_t chan)
+{
+	if (nrf_rtc_int_enable_check(RTC, RTC_CHANNEL_INT_MASK(chan)) &&
+	    nrf_rtc_event_check(RTC, RTC_CHANNEL_EVENT_ADDR(chan))) {
+		uint32_t counter;
+		uint32_t overflow;
+		uint64_t curr_time;
+		z_nrf_rtc_timer_compare_handler_t handler;
+		bool shall_execute_handler = false;
+		void *user_context;
+		uint32_t mcu_critical_state;
+
+		event_clear(chan);
+		event_disable(chan);
+		overflow_and_counter_get(&overflow, &counter);
+
+		curr_time =
+		    overflow_and_counter_to_target_time(overflow, counter);
+
+		/* This critical section is used to provide atomic access to
+		 * cc_data structure and prevent higher priority contexts
+		 * (including ZLIs) from overwriting it.
+		 */
+		mcu_critical_state = __get_PRIMASK();
+		__disable_irq();
+
+		/* If target_cc is in the past or is equal to current counter
+		 * value, execute the handler.
+		 */
+		if (counter_sub(cc_data[chan].target_cc, counter + 1)
+		    > COUNTER_HALF_SPAN) {
+			handler = cc_data[chan].callback;
+			user_context = cc_data[chan].user_context;
+			cc_data[chan].callback = NULL;
+
+			shall_execute_handler = handler != NULL;
```

Comment:
You can give up with separate variable `shall_execute_handler`.
Let's simply have initial value in line 447
```c
z_nrf_rtc_timer_compare_handler_t handler = NULL
```
then
```c
if (handler)
{
    // ... 
    handler()
}
```